### ISSUE TITLE
Refactor CaseParser to use BeautifulSoup with the lxml parser backend

### ISSUE THREAD

**Issue by KentShikama** [codeforpdx/recordexpungPDX#655]
I believe we need to revisit [the decision to not touch the case parser](https://github.com/codeforpdx/recordexpungPDX/pull/556#pullrequestreview-317798777).

- In a meeting yesterday, Michael requested a feature to display relevant practical information relating to the county of the case such as "X county requires a fee waiver form for each individual charge." This will involve modifying the case parser to parse information about the county.
- To properly handle diverted DUIIs, we would need to handle cases where there are multiple disposition rulings corresponding to one charge (e.g, a "diversion" followed by a "dismissed" is a "diverted"). Currently the case parser only returns one disposition ruling per charge, and so this will require a modification. (Edit: Thanks to @NickSchimek's comment below, I realized this isn’t two dispositions as much as a diversion followed by a disposition but in either case it will require that multiple “events” to be used to build up the Disposition, and thus a modification to the parser as we filter out non-disposition events).
- In addition to the above examples, my impression is that there will be more features that require touching the case parser in the future.
- We have recently seen a bug involving the case parser: see https://github.com/codeforpdx/recordexpungPDX/pull/646
- The current handle_start_tag, handle_end_tag pattern with the parser state is hard to follow in my opinion. The default Beautiful Soup pattern of using `find` to fetch out the relevant DOM subtree and handling that at once should be easier to follow.
- It should not be that large of a change. Note we can leave the record, param, and node parsers alone as those do not look like they will change for the time being. The case parser itself is ~130 lines of code and including the parser state files, it is only around ~200 lines. The logic itself isn't that complicated either, and we do have test coverage.
- The case parser is currently our biggest bottleneck along with fetching requests from OECI. The following shows the relevant parts of the [stat dump from the cProfile](https://docs.python.org/3/library/profile.html#profile.Profile) on the search endpoint. Note that I've re-serialized the requests to OECI (otherwise the stat dump gets a bit harder to read), and so the actual total time spent in requests to OECI is lower. Roughly, we spend around 60% in parsing HTML, 30% fetching pages from OECI, 5% on the "fuzzy search" to detect revoked probations, and less than 1% on the time analyzer. Switching to Beautiful Soup with the lxml parser backend should [decrease the time parsing HTML over our current html.parser backend](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#installing-a-parser). Currently for a record of around 100 cases, it takes around ~15 seconds on my localhost. If we have 2+ simultaneous users or a much larger record, it will probably timeout on our current Heroku setup. Firefox will 504 timeout regardless after 60 seconds by default. We can of course rearchitect the request to have the server ping the client when the analysis is ready but I do think that is unnecessary complexity for the time being.
```
         5190202 function calls (5182510 primitive calls) in 81.834 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.005    0.005   81.310   81.310 ./expungeservice/crawler/crawler.py:34(search)
      168    0.013    0.000   51.430    0.306 ./expungeservice/crawler/parsers/case_parser/case_parser.py:30(feed)
      171    0.011    0.000   25.028    0.146 /usr/local/lib/python3.7/dist-packages/requests/sessions.py:466(request)

...

      171    7.746    0.045   49.794    0.291 /usr/lib/python3.7/html/parser.py:134(goahead)
    93287    8.938    0.000   21.888    0.000 /usr/lib/python3.7/html/parser.py:301(parse_starttag)
      203   17.163    0.085   17.163    0.085 {method 'read' of '_ssl._SSLSocket' objects}
    85186    3.779    0.000    9.198    0.000 /usr/lib/python3.7/html/parser.py:386(parse_endtag)
   357526    3.624    0.000    4.953    0.000 /usr/lib/python3.7/_markupbase.py:48(updatepos)
   645180    4.186    0.000    4.186    0.000 {method 'match' of 're.Pattern' objects}
      168    3.912    0.023    4.136    0.025 ./expungeservice/crawler/fuzzy_search.py:7(search)
    93287    1.707    0.000    2.964    0.000 /usr/lib/python3.7/html/parser.py:352(check_for_whole_start_tag)
    89128    1.513    0.000    2.647    0.000 ./expungeservice/crawler/parsers/case_parser/case_parser.py:41(handle_endtag)
    89128    1.430    0.000    2.483    0.000 ./expungeservice/crawler/parsers/case_parser/case_parser.py:34(handle_starttag)
   153995    1.036    0.000    2.093    0.000 /usr/lib/python3.7/html/__init__.py:122(unescape)
   369382    1.756    0.000    1.756    0.000 {method 'end' of 're.Match' objects}
   309077    1.610    0.000    1.610    0.000 {method 'group' of 're.Match' objects}
   318553    1.594    0.000    1.594    0.000 {method 'lower' of 'str' objects}
   275525    1.447    0.000    1.447    0.000 {method 'startswith' of 'str' objects}
    54788    0.729    0.000    1.409    0.000 ./expungeservice/crawler/parsers/case_parser/case_parser.py:59(handle_data)
   237014    1.261    0.000    1.261    0.000 {method 'count' of 'str' objects}
```
- Now is our chance. Michael is still our only real user. It will become much harder to justify a refactor when we have multiple users and the software has a reputation of being stable.

@NickSchimek @wittejm

**Comment by NickSchimek** [comment]
> In a meeting yesterday, Michael requested a feature to display relevant practical information relating to the county of the case such as "X county requires a fee waiver form for each individual charge." This will involve modifying the case parser to parse information about the county.

FYI: This information is available as `case.location`

> The current handle_start_tag, handle_end_tag pattern with the parser state is hard to follow in my opinion.

agreed.

> To properly handle diverted DUIIs, we would need to handle cases where there are multiple disposition rulings corresponding to one charge (e.g, a "convicted" followed by a "dismissed" is a "diverted"). Currently the case parser only returns one disposition ruling per charge, and so this will require a modification.

Can you send me an example of this that I can view. I haven't seen an example of a charge with multiple dispositions.

The main reason for not touching the parser was that at that time we were rolling out a release. Now that the release has been rolled out. I say go ahead, and do whatever you and @wittejm  decide on.

**Comment by KentShikama** [comment]
> FYI: This information is available as case.location

Ah, thanks for the catch. I was too focused on the Case Parser and didn't realize it was injected from the records page.

> Can you send me an example of this that I can view. I haven't seen an example of a charge with multiple dispositions.

Will do.

> I say go ahead, and do whatever you and @wittejm decide on.

Great!

**Comment by wittejm** [comment]
This sounds good to me.

**Comment by NickSchimek** [comment]
@KentShikama in regards to your comment  

> Currently for a record of around 100 cases, it takes around ~15 seconds on my localhost.

I would re-evaluate your time analysis of the case expunger. setting up a simple speed test of the case parser shows that it takes 2.9 seconds to parse 150 cases. The additional 12 seconds that you see when running it live is most likely coming from the 100 get requests that is required. Further work on issue #260 is most likely needed to see a significant improvement in speed.

**Comment by KentShikama** [comment]
> I would re-evaluate your time analysis of the case expunger.

Indeed. I shouldn't have simplified the problem to measure the serialized version as that is measuring something very different.
